python/lsst/daf/persistence/repository.py

```diff
@@ -29,231 +29,100 @@
-        # - None: look for the mapper named in 'access' and use that string as in item 2.
-        mapper = repoCfg['mapper']
-        if mapper is None:
-            if self._access is not None:
-                mapper = self._access.mapperClass()
-            if mapper is None:
-                self._mapper = None
-                return None
-        # if mapper is a cfg (IE an instance of the badly-named Policy class), instantiate via the cfg.
-        if isinstance(mapper, Policy):
-            # code at this location that knows that the mapper needs to share the repo's access instance is
-            # not ideal IMO. Not sure how to rectify in a good way.
-            if mapper['access'] is None:
-                #mapper = copy.copy(mapper)
-                mapper['access'] = self._access
-            mapper = Mapper.Mapper(mapper)
+        # - a class object: instantiate it with mapperArgs
+        mapper = repositoryCfg.mapper
+
         # if mapper is a string, import it:
         if isinstance(mapper, basestring):
             mapper = __import__(mapper)
         # now if mapper is a class type (not instance), instantiate it:
         if inspect.isclass(mapper):
-            # cameraMapper requires root which is not ideal. it should be accessing objects via storage.
-            # cameraMapper and other existing mappers (hscMapper) will require much refactoring to support this.
-            args = inspect.getargspec(mapper.__init__)
-            useRootKeyword = not 'cfg' in args.args
-            if not useRootKeyword:
-                try:
-                    # try new style init first; pass cfg to mapper
-                    mapper = mapper(cfg=repoCfg['mapperCfg'])
-                except TypeError:
-                    # try again, using old style cfg: using mapperArgs and root keywords
-                    useRootKeyword = True
-            if useRootKeyword:
-                mapperArgs = copy.copy(repoCfg['mapperArgs'])
-                if mapperArgs is None:
-                    mapperArgs = {}
-                if 'root' not in mapperArgs:
-                    mapperArgs['root'] = self._access.root()
-                mapper = mapper(**mapperArgs)
+            mapperArgs = copy.copy(repositoryCfg.mapperArgs)
+            if mapperArgs is None:
+                mapperArgs = {}
+            # so that root doesn't have to be redundantly passed in cfgs, if root is specified in the
+            # storage and if it is an argument to the mapper, make sure that it's present in mapperArgs.
+            for arg in ('root', 'storage'):
+                if arg not in mapperArgs:
+                    mro = inspect.getmro(mapper)
+                    if mro[-1] is object:
+                        mro = mro[:-1]
+                    for c in mro:
+                        if arg in inspect.getargspec(c.__init__ ).args:
+                            mapperArgs[arg] = self._storage.root
```

Comment:
eventually the mapper should have no concept of root. we will be working towards that over the next year.